src/Apps/Juno/Command.hs

```diff
@@ -38,80 +43,85 @@ setBalances :: Map.Map Text Rational -> DEval.PersistentState -> DEval.Persisten
         (\e -> do
-            MV.putMVar mvar (ps,ss)
+            MV.putMVar mvar (ps, ss, store)
             return $ BSC.pack $ show (e :: SomeException)) $
         case cmd of
             CreateAccount acct ->
                 if Map.member acct bals
                 then do
-                    MV.putMVar mvar (ps,ss)
+                    MV.putMVar mvar (ps, ss, store)
                     return "Account Already Exists"
                 else do
-                    MV.putMVar mvar $! (setBalances (Map.insert acct 0 bals) ps, ss)
+                    MV.putMVar mvar $! (setBalances (Map.insert acct 0 bals) ps, ss, store)
                     return $ BSC.pack $ "Created Account: " ++ show acct
 
             AdjustAccount acct amount ->
                 if Map.member acct bals
                 then do
-                    MV.putMVar mvar $! (setBalances (Map.adjust (+ amount) acct bals) ps, ss)
+                    MV.putMVar mvar $! (setBalances (Map.adjust (+ amount) acct bals) ps, ss, store)
                     return $ BSC.pack $ "Adjusted Account " ++ show acct ++ " by " ++ show amount
                 else do
-                    MV.putMVar mvar (ps, ss)
+                    MV.putMVar mvar (ps, ss, store)
                     return $ BSC.pack $ "Error: Account " ++ show acct ++ " does not exist!"
 
             ObserveAccount acct -> do
-                MV.putMVar mvar (ps,ss)
+                MV.putMVar mvar (ps, ss, store)
                 return $ BSC.pack $ show $ Map.lookup acct bals
 
             ObserveAccounts -> do
-                MV.putMVar mvar (ps, ss)
+                MV.putMVar mvar (ps, ss, store)
                 return $ BSC.pack $ prettyLedger bals
 
-            Program term ->
+            CommandInputQuery rid' -> do
+                 MV.putMVar mvar (ps, ss, store)
+                 return $ BSC.pack $ show $ Map.lookup rid' store
+
+            Program input term -> do
                 case DTerm.evaluableHopliteTerm term of
                     Nothing -> do
-                      MV.putMVar mvar (ps, ss)
-                      return "Invalid Command or Program"
+                      MV.putMVar mvar (ps, ss, store)
+                      return $  BSC.pack $ "Invalid Command or Program" ++ (show term)
                     Just (DTerm.PolyF evaluableTerm) ->
                         case DEval.runExpr 10000 ps evaluableTerm of
                             Left err -> do
-                                MV.putMVar mvar (ps, ss)
+                                MV.putMVar mvar (ps, ss, store)
                                 return $ BSC.pack $ "Execution issue " ++ show err
                             Right (DEval.InterpreterOutput (DEval.InterpreterDiff _balsDiff ops)
                                                         nextPs) -> do
-                                MV.putMVar mvar (nextPs, ss)
+                                store' <- return $ Map.insert rid input store
```

Comment:
inline store' into the mvar put (avoid unnecc. vars, esp all of these 'tick ones, wherever poss)